banks-server/src/banks_server.rs

```diff
@@ -85,8 +85,21 @@ impl BanksServer {
                 .into_iter()
                 .map(|info| deserialize(&info.wire_transaction).unwrap())
                 .collect();
-            let bank = bank_forks.read().unwrap().working_bank();
-            let _ = bank.try_process_transactions(transactions.iter());
+            'inner: loop {
+                let bank = bank_forks.read().unwrap().working_bank();
+                // bank forks lock released, now verify bank hasn't been frozen yet
+                // in the mean-time the bank can not be frozen until this tx batch
+                // has been processed, this replicates BankingStage behaviour
```

Comment:
this isn't quite how BankingStage does it. We only hold the freeze lock to prevent freezing after execution, while we are recording/committing.

Ref: https://github.com/solana-labs/solana/blob/master/core/src/banking_stage/consumer.rs#L524-L579

Reply:
Because we don't take the lock anywhere inside `try_process_transactions`, I think it's better to grab and hold the lock. @mvines suggestion seems clean (can also remove the loop label)